packages/template-retail-react-app/app/commerce-hooks/context/shopper-basket-context.js

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import React, {useEffect} from 'react'
+import {useCommerceAPI} from '../../commerce-api/contexts'
+import useCustomer from '../../commerce-api/hooks/useCustomer'
+import basketsReducer from '../reducers/baskets-reducer'
+
+const basketsInitialState = {
+    isLoading: false,
+    error: false,
+    data: [],
+    total: 0
+}
+const BasketsContext = React.createContext()
+const BasketsProvider = ({children}) => {
+    // TODO: we should separate api into different apis context instead of having a combined api object here
+    // for example: ShopperBasketAPI, ShopperCustomerAPI
+    // expect to use const api = shopperBasketAPI()
+    const api = useCommerceAPI()
+
+    // slas hooks
+    const shopper = useCustomer()
+    // Create or restore the user session upon mounting
+    // this flow should stay in shopper hook
+    useEffect(() => {
+        shopper.login()
+    }, [])
+
+    const [baskets, dispatch] = React.useReducer(basketsReducer, basketsInitialState)
+
+    const value = React.useMemo(() => ({baskets, dispatch}), [baskets])
+
+    const createBasket = async () => {
+        dispatch({type: 'loading'})
+        try {
+            const res = await api.shopperBaskets.createBasket({})
+            console.log('seting a basket====================')
+            dispatch({type: 'set_basket', payload: res})
+            return res
+        } catch (err) {
+            dispatch({type: 'error'})
+            throw new Error(err)
+        }
+    }
+
+    const getBaskets = async () => {
+        dispatch({type: 'loading'})
+        try {
+            console.log('getting baskets---------------------------')
+            console.log('api.shopperBaskets', api.shopperBaskets)
+            const res = await api.shopperCustomers.getCustomerBaskets({
+                parameters: {customerId: shopper?.customerId}
+            })
+            console.log('res', res)
+            if (!res.total) {
+                console.log('creating a basket ----------------------------')
+                await createBasket()
+            }
+
+            console.log('setting baskets-------------------------')
+            dispatch({type: 'set_baskets', payload: res})
+        } catch (e) {
+            console.log('basket error --------------------')
+            console.log('e', e)
+        }
+    }
+
+    const addItemToCart = async (item, basketId) => {
```

Comment:
Having the action inside the provider might be more straightforwards in terms of interface since they are all inside the context. The users only need to destruct the action from the context
```
import {useBaskets} from 'ecom-hooks`


const PDP = () => {
  const {baskets, addItemToCart} = useBaskets()
  return <div>PDP</div>
}
```